pkg/controller/endpointslice/topologycache/utils.go

```diff
@@ -0,0 +1,150 @@
+/*
+Copyright 2021 The Kubernetes Authors.
+
+Licensed under the Apache License, Version 2.0 (the "License");
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package topologycache
+
+import (
+	"math"
+
+	"k8s.io/api/core/v1"
+	discovery "k8s.io/api/discovery/v1beta1"
+)
+
+// RemoveHintsFromSlices removes topology hints on EndpointSlices and returns
+// updated lists of EndpointSlices to create and update.
+func RemoveHintsFromSlices(si *SliceInfo) ([]*discovery.EndpointSlice, []*discovery.EndpointSlice) {
+	slices := si.ToCreate
+	for _, slice := range si.ToUpdate {
+		slices = append(slices, slice)
+	}
+
+	for _, slice := range slices {
+		for i := range slice.Endpoints {
+			slice.Endpoints[i].Hints = nil
+		}
+	}
+
+	j := 0
+	for _, slice := range si.Unchanged {
+		changed := false
+		for i, endpoint := range slice.Endpoints {
+			if endpoint.Hints != nil {
+				slice = slice.DeepCopy()
```

Comment:
Do we understand why we are deep-copying or is it paranoia?

Reply:
Intentional but needed a comment - added one now. Back in the day Jordan helped me add a cacheMutationCheck util to EndpointSlice controller tests to ensure we weren't modifying resources in informer cache. Unchanged slices are direct references to informer cache until we reach this point.